oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java

```diff
@@ -0,0 +1,175 @@
+import com.linecorp.armeria.common.auth.oauth2.AccessTokenCapsule;
+import com.linecorp.armeria.common.auth.oauth2.ClientAuthorization;
+import com.linecorp.armeria.common.auth.oauth2.RefreshAccessTokenRequest;
+
+abstract class AbstractOAuth2AuthorizationGrantBuilder {
+
+    private final WebClient accessTokenEndpoint;
+    private final String accessTokenEndpointPath;
+
+    @Nullable
+    private ClientAuthorization clientAuthorization;
+
+    @Nullable
+    private Duration refreshBefore;
+
+    @Nullable
+    private Supplier<AccessTokenCapsule> tokenSupplier;
+
+    @Nullable
+    private Consumer<AccessTokenCapsule> tokenConsumer;
+
+    /**
+     * A common abstraction for the requests implementing various Access Token request/response flows,
+     * as per <a href="https://tools.ietf.org/html/rfc6749">[RFC6749]</a>.
+     *
+     * @param accessTokenEndpoint A {@link WebClient} to facilitate an Access Token request. Must correspond to
+     *                            the Access Token endpoint of the OAuth 2 system.
+     * @param accessTokenEndpointPath A URI path that corresponds to the Access Token endpoint of the
+     *                                OAuth 2 system.
+     */
+    AbstractOAuth2AuthorizationGrantBuilder(WebClient accessTokenEndpoint, String accessTokenEndpointPath) {
+        this.accessTokenEndpoint = requireNonNull(accessTokenEndpoint, "accessTokenEndpoint");
+        this.accessTokenEndpointPath = requireNonNull(accessTokenEndpointPath, "accessTokenEndpointPath");
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on encoded authorization token and
+     * authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param authorizationSupplier A supplier of encoded client authorization token.
+     * @param authorizationType One of the registered HTTP authentication schemes as per
+     *                          <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">
+     *                          HTTP Authentication Scheme Registry</a>.
+     */
+    protected AbstractOAuth2AuthorizationGrantBuilder clientAuthorization(
+            Supplier<String> authorizationSupplier, String authorizationType) {
+        clientAuthorization = ClientAuthorization.ofAuthorization(authorizationSupplier, authorizationType);
+        return this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on encoded authorization token and
+     * {@code Basic} authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param authorizationSupplier A supplier of encoded client authorization token.
+     */
+    protected AbstractOAuth2AuthorizationGrantBuilder clientBasicAuthorization(
+            Supplier<String> authorizationSupplier) {
```

Comment:
How about making all these protected setters package-local? Looks like all subclasses are in the same package.

Reply:
made those methods public and used different builder inheritance technique.
btw, it's expected to have more grants and grant builders in future.